Jaqueline Pinheiro

In [143]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import scipy.stats as stats
from scipy.stats import levene
import statsmodels.api as sm

## Question 1
### letter d)


In [144]:
df = pd.read_excel('SP500Index.xlsx')
df

Date of Observation  Level of the S&P 500 Index
0               19600129                       55.61
1               19600229                       56.12
2               19600331                       55.34
3               19600429                       54.37
4               19600531                       55.83
..                   ...                         ...
697             20180228                     2713.83
698             20180329                     2640.87
699             20180430                     2648.05
700             20180531                     2705.27
701             20180629                     2718.37

[702 rows x 2 columns]

In [145]:
df['Date of Observation'] = pd.to_datetime(df['Date of Observation'], format='%Y%m%d')
df['xt'] = np.log(df['Level of the S&P 500 Index']) - np.log(df['Level of the S&P 500 Index'].iloc[0])
mean_return=df.iloc[-1].xt/len(df)
mean_return

0.005540490988253315

In [146]:
std = np.sqrt(((1/len(df))*(df['xt'] - df['xt'].shift(1) - delta)**2).sum())
std

0.04215029312565144

In [147]:
std/mean_return

7.607681921154007

<br>$\hat{\delta}_{M L}$ = 0.00554 (average monthly return on SP500 index between Jan 1960 - Jun 2018)
<br>$\sqrt{\hat{\sigma}_{M L}^{2}}$ = 0.04215 (volatility of returns on SP500 index between Jan 1960 - Jun 2018)

The volatility of the index is much bigger than its return.

## Question 5
### letter a)

In [148]:
df = pd.read_excel('PS4data.xls')
df

year  quarter  real consumption of nondurables  \
0    1948        1                            442.3   
1    1948        2                            446.9   
2    1948        3                            443.3   
3    1948        4                            449.6   
4    1949        1                            451.8   
..    ...      ...                              ...   
215  2001        4                           1885.0   
216  2002        1                           1921.4   
217  2002        2                           1920.9   
218  2002        3                           1925.8   
219  2002        4                           1950.0   

     real consumption of services         cpi  real disposable income  \
0                           463.8   23.533333                  1055.3   
1                           470.5   23.933333                  1087.7   
2                           476.6   24.466667                  1107.1   
3                           480.2   24.233333                  1109.8   
4                           482.6   23.866667                  1087.8   
..                            ...         ...                     ...   
215                        3616.6  177.266670                  6729.1   
216                        3642.2  177.900000                  6961.0   
217                        3666.2  179.833330                  7027.2   
218                        3687.0  180.600000                  7058.1   
219                        3707.0  181.166670                  7100.8   

     population  \
0     102690.67   
1     102915.33   
2     103249.00   
3     103417.67   
4     103584.33   
..          ...   
215   216111.67   
216   216664.00   
217   217203.67   
218   217867.67   
219   218543.00   

     equally weighted average return on the New York Stock Exchange  
0                                              0.01414               
1                                              0.11755               
2                                             -0.09565               
3                                             -0.04563               
4                                              0.03338               
..                                                 ...               
215                                            0.14591               
216                                            0.08077               
217                                           -0.03911               
218                                           -0.16290               
219                                            0.07456               

[220 rows x 8 columns]

In [149]:
df['C0'] = (df['real consumption of nondurables'] + df['real consumption of services'])/df['population']
df['C1'] = df['C0'].shift(1)
df['C2'] = df['C1'].shift(1)
df['C3'] = df['C2'].shift(1)
df['C4'] = df['C3'].shift(1)
df = df.dropna()
df

year  quarter  real consumption of nondurables  \
4    1949        1                            451.8   
5    1949        2                            452.7   
6    1949        3                            449.2   
7    1949        4                            455.5   
8    1950        1                            461.1   
..    ...      ...                              ...   
215  2001        4                           1885.0   
216  2002        1                           1921.4   
217  2002        2                           1920.9   
218  2002        3                           1925.8   
219  2002        4                           1950.0   

     real consumption of services         cpi  real disposable income  \
4                           482.6   23.866667                  1087.8   
5                           485.7   23.866667                  1091.3   
6                           484.5   23.800000                  1096.8   
7                           486.0   23.700000                  1106.3   
8                           493.6   23.533333                  1186.1   
..                            ...         ...                     ...   
215                        3616.6  177.266670                  6729.1   
216                        3642.2  177.900000                  6961.0   
217                        3666.2  179.833330                  7027.2   
218                        3687.0  180.600000                  7058.1   
219                        3707.0  181.166670                  7100.8   

     population  \
4     103584.33   
5     103838.00   
6     104127.33   
7     104427.67   
8     104733.33   
..          ...   
215   216111.67   
216   216664.00   
217   217203.67   
218   217867.67   
219   218543.00   

     equally weighted average return on the New York Stock Exchange        C0  \
4                                              0.03338               0.009021   
5                                             -0.08055               0.009037   
6                                              0.13730               0.008967   
7                                              0.12362               0.009016   
8                                              0.05450               0.009116   
..                                                 ...                    ...   
215                                            0.14591               0.025457   
216                                            0.08077               0.025678   
217                                           -0.03911               0.025723   
218                                           -0.16290               0.025762   
219                                            0.07456               0.025885   

           C1        C2        C3        C4  
4    0.008991  0.008910  0.008914  0.008824  
5    0.009021  0.008991  0.008910  0.008914  
6    0.009037  0.009021  0.008991  0.008910  
7    0.008967  0.009037  0.009021  0.008991  
8    0.009016  0.008967  0.009037  0.009021  
..        ...       ...       ...       ...  
215  0.025373  0.025387  0.025409  0.025397  
216  0.025457  0.025373  0.025387  0.025409  
217  0.025678  0.025457  0.025373  0.025387  
218  0.025723  0.025678  0.025457  0.025373  
219  0.025762  0.025723  0.025678  0.025457  

[216 rows x 13 columns]

In [151]:
# Linear model 
clist = df.columns[-4:].tolist()
X = df[clist]
X = sm.add_constant(X)

Y = df[df.columns[-5]]

linear_model = sm.OLS(Y,X)
model_1 = linear_model.fit()
print(model_1.summary()) 

                            OLS Regression Results                            
Dep. Variable:                     C0   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 2.094e+05
Date:                Wed, 22 Nov 2023   Prob (F-statistic):               0.00
Time:                        20:51:15   Log-Likelihood:                 1745.1
No. Observations:                 216   AIC:                            -3480.
Df Residuals:                     211   BIC:                            -3463.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.439e-05   1.92e-05      1.267      0.2

In [152]:
## Runing hypo. test
hypotheses = '(C2 = 0), (C3 = 0), (C4 = 0)'
f_test = model_1 .f_test(hypotheses)
print(f_test)

<F test: F=8.40999689552861, p=2.6364145736173065e-05, df_denom=211, df_num=3>


<br> The F test rejects the hypothesis that $\beta_2=\beta_3=\beta_4=0$.
<br> We reject the null hypothesis for the first and the fourth lags of consumption at the 1\% sig. level. On the teh other hand, we fail to reject the null hypothesis for lag 2, and 3. The significance of the first lag is consistent with Hall's hypothesis. The significance of the fourth lag is not consistent with his theory, and one can argue that it could be related to seasonality. 

### letter b)

Changes in current income $(Y_t-Y_{t-1})$ are likely correlated with $\epsilon_t$, which provides information about changes in the permanent income. Thus, $(Y_t-Y_{t-1})$ is endogenous and the estimation of $\gamma_1$ is inconsistent. 

### letter c)

In [100]:
import warnings
warnings.filterwarnings("ignore")

In [153]:
df['C5'] = df['C4'].shift(1)
df['C6'] = df['C5'].shift(1)
df['log_c0/c1'] = np.log(df['C0']/df['C1'])
df['log_c2/c3'] = np.log(df['C2']/df['C3'])
df['log_c3/c4'] = np.log(df['C3']/df['C4'])
df['log_c4/c5'] = np.log(df['C4']/df['C5'])
df['log_c5/c6'] = np.log(df['C5']/df['C6'])
df['Y0'] = df['real disposable income']/df['population']
df['Y1'] = df['Y0'].shift(1)
df = df.dropna()
df['log_y0/y1'] = np.log(df['Y0']/df['Y1'])
df

year  quarter  real consumption of nondurables  \
6    1949        3                            449.2   
7    1949        4                            455.5   
8    1950        1                            461.1   
9    1950        2                            466.1   
10   1950        3                            474.4   
..    ...      ...                              ...   
215  2001        4                           1885.0   
216  2002        1                           1921.4   
217  2002        2                           1920.9   
218  2002        3                           1925.8   
219  2002        4                           1950.0   

     real consumption of services         cpi  real disposable income  \
6                           484.5   23.800000                  1096.8   
7                           486.0   23.700000                  1106.3   
8                           493.6   23.533333                  1186.1   
9                           507.6   23.700000                  1178.1   
10                          514.6   24.266667                  1196.5   
..                            ...         ...                     ...   
215                        3616.6  177.266670                  6729.1   
216                        3642.2  177.900000                  6961.0   
217                        3666.2  179.833330                  7027.2   
218                        3687.0  180.600000                  7058.1   
219                        3707.0  181.166670                  7100.8   

     population  \
6     104127.33   
7     104427.67   
8     104733.33   
9     105020.33   
10    105248.33   
..          ...   
215   216111.67   
216   216664.00   
217   217203.67   
218   217867.67   
219   218543.00   

     equally weighted average return on the New York Stock Exchange        C0  \
6                                              0.13730               0.008967   
7                                              0.12362               0.009016   
8                                              0.05450               0.009116   
9                                              0.00104               0.009272   
10                                             0.16630               0.009397   
..                                                 ...                    ...   
215                                            0.14591               0.025457   
216                                            0.08077               0.025678   
217                                           -0.03911               0.025723   
218                                           -0.16290               0.025762   
219                                            0.07456               0.025885   

           C1  ...        C5        C6  log_c0/c1  log_c2/c3  log_c3/c4  \
6    0.009037  ...  0.008914  0.008824  -0.007804   0.003325   0.009072   
7    0.008967  ...  0.008910  0.008914   0.005439   0.001826   0.003325   
8    0.009016  ...  0.008991  0.008910   0.011000  -0.007804   0.001826   
9    0.009116  ...  0.009021  0.008991   0.016970   0.005439  -0.007804   
10   0.009272  ...  0.009037  0.009021   0.013422   0.011000   0.005439   
..        ...  ...       ...       ...        ...        ...        ...   
215  0.025373  ...  0.025280  0.025159   0.003331  -0.000830   0.000472   
216  0.025457  ...  0.025397  0.025280   0.008654  -0.000588  -0.000830   
217  0.025678  ...  0.025409  0.025397   0.001727   0.003331  -0.000588   
218  0.025723  ...  0.025387  0.025409   0.001537   0.008654   0.003331   
219  0.025762  ...  0.025373  0.025387   0.004749   0.001727   0.008654   

     log_c4/c5  log_c5/c6        Y0        Y1  log_y0/y1  
6    -0.000516   0.010209  0.010533  0.010510   0.002245  
7     0.009072  -0.000516  0.010594  0.010533   0.005744  
8     0.003325   0.009072  0.011325  0.010594   0.066727  
9     0.001826   0.003325  0.011218  0.011325  -0.009504  
10   -0.007804   0.001826  0.011368  0.011218   0.013329  
..        

In [154]:
# First Stage
clist = ['log_c2/c3', 'log_c3/c4', 'log_c4/c5', 'log_c5/c6']
X = df[clist]
X = sm.add_constant(X)

Y = df['log_y0/y1'] 

first_stage_model = sm.OLS(Y,X)
model_fs = first_stage_model.fit(cov_type = 'HC0')
print(model_fs.summary()) 

                            OLS Regression Results                            
Dep. Variable:              log_y0/y1   R-squared:                       0.046
Model:                            OLS   Adj. R-squared:                  0.028
Method:                 Least Squares   F-statistic:                     3.208
Date:                Wed, 22 Nov 2023   Prob (F-statistic):             0.0139
Time:                        20:51:56   Log-Likelihood:                 687.27
No. Observations:                 214   AIC:                            -1365.
Df Residuals:                     209   BIC:                            -1348.
Df Model:                           4                                         
Covariance Type:                  HC0                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0040      0.002      2.407      0.0

In [155]:
# Second Stage
X = model_fs.fittedvalues.tolist()
X = sm.add_constant(X)

Y = df['log_c0/c1']

second_stage_model = sm.OLS(Y,X)
model_ss = second_stage_model.fit(cov_type = 'HC0')
print(model_ss.summary()) 

                            OLS Regression Results                            
Dep. Variable:              log_c0/c1   R-squared:                       0.032
Model:                            OLS   Adj. R-squared:                  0.028
Method:                 Least Squares   F-statistic:                     4.638
Date:                Wed, 22 Nov 2023   Prob (F-statistic):             0.0324
Time:                        20:52:07   Log-Likelihood:                 825.70
No. Observations:                 214   AIC:                            -1647.
Df Residuals:                     212   BIC:                            -1641.
Df Model:                           1                                         
Covariance Type:                  HC0                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0026      0.001      2.179      0.0

To address the endogeneity problem, we follow the proposal of using the $\log \left(C_{t-2} / C_{t-3}\right), \log \left(C_{t-3} / C_{t-4}\right), \log \left(C_{t-4} / C_{t-5}\right)$, and $\log \left(C_{t-5} / C_{t-6}\right)$ as instruments, and run the 2SLS. 
<br> From the results, we see that the coefficient is statistically different from 0 at 5\% sig level, which means that 43.7\% of the households are keynesian (i.e. consume their current disposable income directly) instead of consuming according to the permanent income hypothesis.

### letter d)

In [156]:
# Control Function
control = model_fs.resid.tolist()
df['control'] = control

clist = ['log_y0/y1', 'control']
X = df[clist]
X = sm.add_constant(X)

control_model = sm.OLS(Y,X)
model_cf = control_model.fit(cov_type = 'HC0')
print(model_cf.summary()) 

                            OLS Regression Results                            
Dep. Variable:              log_c0/c1   R-squared:                       0.219
Model:                            OLS   Adj. R-squared:                  0.212
Method:                 Least Squares   F-statistic:                     18.33
Date:                Wed, 22 Nov 2023   Prob (F-statistic):           4.56e-08
Time:                        21:03:31   Log-Likelihood:                 848.67
No. Observations:                 214   AIC:                            -1691.
Df Residuals:                     211   BIC:                            -1681.
Df Model:                           2                                         
Covariance Type:                  HC0                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0026      0.001      2.384      0.0

We cannot reject the null hypothesis that $\beta_2=0$ (control variable in our model). Therefore, we cannot reject the hypothesis of exogeneity.

In [157]:
#### Overidentification test

# instead of taking the residuals of the second stage, we recompute using the endogenous variable

res = df['log_c0/c1'] -  model_ss.predict(sm.add_constant(df['log_y0/y1']))

clist = ['log_c2/c3', 'log_c3/c4', 'log_c4/c5', 'log_c5/c6']
X = df[clist]
X = sm.add_constant(X)

over_ident_model = sm.OLS(res,X)
model_over = over_ident_model.fit()
print(model_over.summary()) 

214*model_over.rsquared

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.036
Model:                            OLS   Adj. R-squared:                  0.018
Method:                 Least Squares   F-statistic:                     1.960
Date:                Wed, 22 Nov 2023   Prob (F-statistic):              0.102
Time:                        21:03:33   Log-Likelihood:                 833.72
No. Observations:                 214   AIC:                            -1657.
Df Residuals:                     209   BIC:                            -1641.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0007      0.001     -1.134      0.2

7.7382792292901454

To test for over-identification, we recompute residuals using the endogenous variable (variable res in our model above) and regress it on the instruments to get the $R^2$.
<br> The Sargan's test statistic is: $S= nR^2 = 214*0.036=7.738$
<br> The asymptotic dist under $H_0$ (valid over-identifying restrictions) is $\chi^2(l-k)=\chi^2(4-1)$, that has critical value is $7.815$ at 5% of sig. level. Therefore, we cannot reject the null hypothesis of valid over-identifying restrictions. 

<br> Note that the F-stat from the first stage is $3.208 <10$, which suggests that it is a weak instrument according to Stock-Yogo. So it can be tha case that testing for endogeneity and the over-identifying restrictions may be flawed. 